source/fontloader.ts

```diff
@@ -0,0 +1,227 @@
+            parseFloat(pairs.get('scaleH')!),
+        ];
+    }
+
+    protected handlePage(
+        stream: Array<string>, fontFace: FontFace, filename: string, context: Context): Promise<void> {
+
+        const pairs = this.readKeyValuePairs(stream, ['file']);
+
+        const path = directoryPath(filename);
+        const file = stripped(pairs.get('file')!, ['"', '\r']);
+
+        const pngPath: string = path + file.split('.')[0] + '.png';
+
+        return fontFace.glyphTexture.load(pngPath);
+    }
+
+    protected handleChar(stream: Array<string>, fontFace: FontFace): void {
+        const pairs = this.readKeyValuePairs(stream,
+            ['id', 'x', 'y', 'width', 'height', 'xoffset', 'yoffset', 'xadvance']);
+
+        const index: number = parseInt(pairs.get('id')!, 10);
+        assert(index > 0, `handleChar: glyph index is not greater than 0.`);
+
+        const glyph = new Glyph();
+
+        glyph.index = index;
+
+        const extentScale: GLfloat2 = [
+            1.0 / fontFace.glyphTextureExtent[0],
+            1.0 / fontFace.glyphTextureExtent[1],
+        ];
+        const extent: GLfloat2 = [
+            parseFloat(pairs.get('width')!),
+            parseFloat(pairs.get('height')!),
+        ];
+
+        glyph.subTextureOrigin = [
+            parseFloat(pairs.get('x')!) * extentScale[0],
+            1.0 - (parseFloat(pairs.get('y')!) + extent[1]) * extentScale[1],
+        ];
+
+        glyph.extent = extent;
+
+        glyph.subTextureExtent[0] = extent[0] * extentScale[0];
+        glyph.subTextureExtent[1] = extent[1] * extentScale[1];
+
+        glyph.setBearing(fontFace.base,
+            parseFloat(pairs.get('xoffset')!),
+            parseFloat(pairs.get('yoffset')!),
+        );
+
+        glyph.advance = parseFloat(pairs.get('xadvance')!);
+
+        fontFace.addGlyph(glyph);
+    }
+
+    protected handleKerning(stream: Array<string>, fontFace: FontFace): void {
+        const pairs = this.readKeyValuePairs(stream, ['first', 'second', 'amount']);
+
+        const first: number = parseInt(pairs.get('first')!, 10);
+        assert(first > 0, `handleKerning: first is not greater than 0.`);
+
+        const second: number = parseInt(pairs.get('second')!, 10);
+        assert(second > 0, `handleKerning: second is not greater than 0.`);
+
+        const kerning: number = parseFloat(pairs.get('amount')!);
+
+        fontFace.setKerning(first, second, kerning);
+    }
+
+    protected readKeyValuePairs(stream: Array<string>, mandatoryKeys: Array<string>): StringPairs {
+
+        let key: string;
+        let value: string;
+
+        const pairs: StringPairs = new Map<string, string>();
+
+        for (const s of stream) {
+            const pair = s.split('=');
+            key = pair[0];
+            value = pair[1];
+            pairs.set(key, value);
+        }
+
+        // check if all required keys are provided
+        let valid = true;
+        for (const mandatoryKey of mandatoryKeys) {
```

Comment:
do you find this more readable? or is there actually a difference in the execution?

Reply:
I understand, good point.
I implemented your suggestions. Since there are not so much mandatory keys, I don't think an early-out would noticeably improve performance, so I chose the "more idiomatic javascript".